### DISCUSSION TITLE
Confused about tokio::select!

### DISCUSSION THREAD

**Discussion by FH0** [tokio-rs/tokio#3707] [Q&A] [answered]
The example https://github.com/tokio-rs/tokio/blob/master/examples/chat.rs#L203 shows me that `tokio::select!` is placed in the loop of `loop`. But if branches are ready at the same time, some of them will be canceled.

To ensure all the data can be handled, I have to do this.
```rust
tokio::select! {
    _ = async {
        loop {
            // some works
        }
    } => {}
    _ = async {
        loop {
            // some works
        }
    } => {}
}
```
Is it correct?

**Comment by Darksonn** [comment, marked as answer]
The `tokio::select!` runs the branches you provide it concurrently, but not in parallel. This means that only one branch can do work at any time, and that the macro will alternate between doing work on the branches. Because of this non-parallel approach, it can't happen that two branches become ready at the same time, so e.g. in the linked example, it is not possible to lose a line from the peer because you got a message from `peer.rx` at the same time.

**Reply by OverShifted** [reply]
> the macro will alternate between doing work on the branches.

So it might switch between tasks without an `await`?

**Reply by Noah-Kennedy** [reply]
Think of it as being as though the select has a hidden await inside of it.
